File: python/beeai_framework/runnable.py
# Copyright 2025 © BeeAI a Series of LF Projects, LLC
# SPDX-License-Identifier: Apache-2.0

import functools
from abc import ABC, abstractmethod
from collections.abc import Awaitable, Callable
from typing import Any, TypedDict, Unpack

from pydantic import BaseModel
from typing_extensions import ParamSpec, TypeVar

from beeai_framework.backend import AnyMessage
from beeai_framework.context import Run, RunContext, RunMiddlewareType
from beeai_framework.emitter import Emitter
from beeai_framework.utils import AbortSignal
from beeai_framework.utils.dicts import exclude_keys

T = TypeVar("T")
P = ParamSpec("P")


class RunnableOptions(TypedDict, total=False):
    """Options for a runnable."""

    signal: AbortSignal
    """The runnable's abort signal data"""

    context: dict[str, Any]
    """Context can be used to pass additional context to runnable"""


class RunnableOutput(BaseModel):
    """Runnable output."""

    output: list[AnyMessage]
    """The runnable output"""

    context: dict[str, Any] | None = None
    """Context can be used to return additional data by runnable"""


class Runnable(ABC):
    """A unit of work that can be invoked using a stable interface.

    For convenience, concrete subclasses can decorate `run` with `runnable_entry`
    to automatically wrap the runnable into an execution context:

        @runnable_entry
        async def run(self, input: str | list[AnyMessage], /, **kwargs: Unpack[AgentOptions]) -> RunnableOutput:
            # ... agent logic ...
            return AgentOutput(output=...)

    Attributes:
        _middleware: The middleware to be used when executing the runnable.
    """

    def __init__(self, middleware: list[RunMiddlewareType] | None = None) -> None:
        super().__init__()
        self._middleware = middleware or []

    @abstractmethod
    def run(self, input: list[AnyMessage], /, **kwargs: Unpack[RunnableOptions]) -> Run[RunnableOutput]:
        """ "Execute the runnable.

        Args:
            input: The input to the runnable
            signal: The runnable abort signal
            context: A dictionary that can be used to pass additional context to the runnable

        Returns:
            The runnable output.
        """
        pass

    @property
    @abstractmethod
    def emitter(self) -> Emitter:
        """The event emitter for the runnable."""
        pass

    @property
    def middleware(self) -> list[RunMiddlewareType]:
        """The middleware to be used when executing the runnable."""
        return self._middleware


def runnable_entry(handler: Callable[P, Awaitable[T]]) -> Callable[P, Run[T]]:
    """A decorator that wraps the runnable into an execution context."""

    @functools.wraps(handler)
    def wrapper(*args: P.args, **kwargs: P.kwargs) -> Run[T]:
        """Wrapper that automates the call to RunContext.enter()."""

        async def inner(_: RunContext) -> T:
            return await handler(*args, **kwargs)

        self = args[0] if args else None
        if not isinstance(self, Runnable):
            raise TypeError("The first argument of a runnable must be a Runnable instance.")

        runnable_kwargs: RunnableOptions = kwargs  # type: ignore
        return RunContext.enter(
            self,
            inner,
            signal=runnable_kwargs.get("signal", None),
            run_params={"input": args[1], **exclude_keys(kwargs, {"signal", "input"})},
        ).middleware(*self.middleware)

    return wrapper

File: typescript/src/runnable.ts
/**
 * Copyright 2025 © BeeAI a Series of LF Projects, LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { Emitter } from "@/emitter/emitter.js";
import { Message, AssistantMessage } from "@/backend/message.js";

// TS exports the message union type as `Message` (Python calls the
// equivalent `AnyMessage`). Alias it locally to keep this file's naming
// consistent with the Python reference implementation.
type AnyMessage = Message;
import { GetRunContext, MiddlewareType, Run, RunContext, RunInstance } from "@/context.js";

/**
 * Options accepted by a Runnable's `run` method.
 *
 * Mirrors Python's `RunnableOptions` (TypedDict).
 */
export interface RunnableOptions {
  /** Abort signal for cancelling the run. */
  signal?: AbortSignal;
  /** Additional context to pass into the runnable. */
  context?: Record<string, any>;
}

/**
 * Output produced by a Runnable.
 *
 * Mirrors Python's `RunnableOutput` (pydantic BaseModel with `extra="allow"`).
 */
export class RunnableOutput {
  /** The runnable output messages. */
  public output: AnyMessage[];

  /** Additional context returned by the runnable. */
  public context: Record<string, any>;

  // Allow arbitrary extra fields, mirroring Python's `extra="allow"`.
  [key: string]: any;

  constructor(input: { output: AnyMessage[]; context?: Record<string, any> } & Record<string, any>) {
    const { output, context, ...extra } = input;
    this.output = output;
    this.context = context ?? {};
    Object.assign(this, extra);
  }

  /**
   * Returns the latest message in output, with a fallback if it is not defined.
   */
  get lastMessage(): AnyMessage {
    const last = this.output.at(-1);
    return last ?? new AssistantMessage("");
  }
}

/**
 * A unit of work that can be invoked using a stable interface.
 *
 * Mirrors Python's `Runnable[R]` ABC. Concrete subclasses (agents, workflows,
 * chat models, ...) implement `run` using `RunContext.enter`, matching the
 * pattern already used elsewhere in this codebase (see `@/context.js`).
 */
export abstract class Runnable<R extends RunnableOutput = RunnableOutput> implements RunInstance {
  protected readonly _middlewares: MiddlewareType<this>[];

  constructor(middlewares: MiddlewareType<any>[] = []) {
    this._middlewares = middlewares as MiddlewareType<this>[];
  }

  /**
   * Execute the runnable.
   *
   * @param input The input messages to the runnable.
   * @param options Execution options (signal, context).
   */
  abstract run(input: AnyMessage[], options?: RunnableOptions): Run<R, this>;

  /** The event emitter for the runnable. */
  abstract get emitter(): Emitter<any>;

  /** The list of middleware to be used when executing the runnable. */
  get middlewares(): MiddlewareType<this>[] {
    return this._middlewares;
  }
}

export type AnyRunnable = Runnable<any>;

/**
 * Helper to wrap a runnable's handler logic into a `RunContext`-managed execution.
 *
 * This is the TypeScript equivalent of Python's `runnable_entry` decorator.
 * Python relies on a decorator to auto-wrap `run()` with `RunContext.enter()`;
 * TS decorators are still experimental/inconsistent across build tools, so
 * this is implemented as a plain helper function instead, called directly
 * from within a concrete `run()` implementation:
 *
 * ```ts
 * class MyRunnable extends Runnable<RunnableOutput> {
 *   run(input: AnyMessage[], options?: RunnableOptions) {
 *     return runnableEntry(this, input, options, async (ctx, input) => {
 *       // `input` here reflects any middleware modifications to ctx.runParams.
 *       // ... runnable logic using ctx / input ...
 *       return new RunnableOutput({ output: [...] });
 *     });
 *   }
 *
 *   get emitter() {
 *     return this._emitter;
 *   }
 * }
 * ```
 */
export function runnableEntry<I extends Runnable<R>, R extends RunnableOutput, M extends AnyMessage = AnyMessage>(
  instance: I,
  input: M[],
  options: RunnableOptions | undefined,
  handler: (context: GetRunContext<I>, input: M[]) => Promise<R>,
): Run<R, I> {
  return RunContext.enter(
    instance,
    { params: input, signal: options?.signal },
    // Read the (possibly middleware-modified) input from `context.runParams`
    // rather than closing over the original `input` argument, so that
    // middleware mutating runParams before execution actually takes effect.
    // This matches Python's `runnable_entry`, which reads
    // `ctx.run_params.get("input", ...)` for the same reason.
    (context) => handler(context, context.runParams ?? input),
  )
    .middleware(...instance.middlewares)
    .context(options?.context ?? {});
}